File: Julia/algorithmic_toolbox/15-maximizing-salary.jl
largestfrompieces(numbers) = join(sort(numbers; lt=(n1, n2) -> n1 * n2 >= n2 * n1))


function main()
    _ = readline()
    numbers = readline() |> split
    result = largestfrompieces(numbers)
    println(result)
end


main()

File: Rust/algorithmic_toolbox/src/bin/15_maximizing_salary.rs
use std::io;

fn read_int() -> i32 {
    let mut x = String::new();
    io::stdin().read_line(&mut x).unwrap();
    x.trim().parse().unwrap()
}

fn read_line() -> Vec<String> {
    let mut line = String::new();
    io::stdin().read_line(&mut line).unwrap();
    line.split_whitespace().map(|x| x.to_string()).collect()
}

fn find_largest_number_from_pieces(mut numbers: Vec<String>) -> String {
    numbers.sort_by(|n, m| {
        let nm = format!("{n}{m}");
        let mn = format!("{m}{n}");
        mn.cmp(&nm)
    });
    numbers.concat()
}

fn main() {
    let _ = read_int();
    let numbers = read_line();
    let result = find_largest_number_from_pieces(numbers);
    println!("{}", result);
}
